### ISSUE TITLE
Tweakreg memory usage improvements

### ISSUE THREAD

**Issue by stscijgbot-jp** [spacetelescope/jwst#8434]
_Issue [JP-3603](https://jira.stsci.edu/browse/JP-3603) was created on JIRA by [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva):_

Refactor tweakreg to improve memory usage. Document improvements.

**Comment by stscijgbot-jp** [comment]
_Comment by [Brett Graham](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bgraham) on [JIRA](https://jira.stsci.edu/browse/JP-3603?focusedCommentId=695469#comment-695469):_

Aiming to improve memory performance for `tweakreg` several changes were identified and implemented in <https://github.com/spacetelescope/jwst/pull/8424:>
- Refactor how custom catalogs are resolved (the behavior should be the same although at the the moment there appear to be no tests for this feature, this PR adds tests). This will allow removal of special handling of `tweakreg_catalog` in `ModelContainer` (one of the reasons that class opens all models when constructed).
- Refactor catalog/corrector and model linking to instead use the catalog/corrector index and model index. These previously were linked by reference and as `align_wcs` requires that all correctors be in-memory this results in all input image_models also required to be in memory. Breaking this link will (eventually) allow the image_models to be loaded one-by-one instead of having them all in memory at the same time.

The PR includes additional (more minor) changes the major changes will be described in more detail below.

Prior to the PR handling of `tweakreg_catalog` entries in the association input file involved opening (and potentially re-saving) all input files so that their `meta.tweakreg_catalog` keywords could be overwritten. The `tweakreg` step was refactored to instead generate a "catalog dictionary" that contained either entries from the association or `meta.tweakreg_catalog` (or one of the numerous other ways a catalog can be provided) removing one need for `ModelContainer` to open files. Additional changes will be needed to remove the `tweakreg_catalog` special handling in `ModelContainer` similar to the special handling needed for `group_id`. Specifically, the container (or a replacement class) can, instead of opening and resaving each input, patch the model with whatever is in the association when the model is opened. As the `group_id` changes will impact other steps, this change was left out of the `tweakreg` PR (and will be in a follow-up PR).

The `tweakreg` step prior to the above PR generated wcs corrector objects (one for each input image) and added a reference to the input image model to the metadata of each corrector object. As `tweakwcs` requires that all correctors be in memory at the time `align_wcs` is called, this meant that the `tweakreg` step was also requiring that all image models also be in memory. This PR breaks this link and instead relies on the index of each corrector and model (they are equal in number of ordered identically). By removing the model reference from the corrector metadata the very high level memory overview of the step is as follows in 3 sub-steps:
- read each image model (incrementally) and construct a wcs corrector object (linked by index to the model)
- run `align_wcs` with all wcs corrector objects in memory
- iterate though each image model and update the wcs based on the output of `align_wcs`
So the total memory load should be as follows for the above 3 sub-steps:
- at each iteration through the image models 1 model will be in memory and a list of correctors (in the end equaling the length of the input image models) is accumulated
- when `align_wcs` is called, a corrector for each image model will be in memory (these are relatively small, estimates are <1MB and having all of these in memory is not a major limitation)
- as updating the input model wcses can occur one at a time, at each iteration in this sub-step 1 model can be in memory

It should be noted that the above memory improvements will require further changes to `ModelContainer`, `stpipe` and the `calwebb_image3` pipeline. Each of these (for different reasons) opens and keeps all models in memory. For example, `calwebb_image3` opens all input image models and passes them in memory to `tweakreg` making the memory improvements outlined above moot (as the models are already all in memory).

**Comment by stscijgbot-jp** [comment]
_Comment by [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) on [JIRA](https://jira.stsci.edu/browse/JP-3603?focusedCommentId=699865#comment-699865):_

Fixed by #8424

**Comment by stscijgbot-jp** [comment]
_Comment by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw) on [JIRA](https://jira.stsci.edu/browse/JP-3603?focusedCommentId=703410#comment-703410):_

Not sure I really know how to test this effectively, so [Howard Bushouse](https://jira.stsci.edu/secure/ViewProfile.jspa?name=bushouse) or [Nadia Dencheva](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dencheva) please close if you're happy with it.

**Comment by stscijgbot-jp** [comment]
_Comment by [Rachel Plesha](https://jira.stsci.edu/secure/ViewProfile.jspa?name=rplesha) on [JIRA](https://jira.stsci.edu/browse/JP-3603?focusedCommentId=710101#comment-710101):_

From [Kevin Volk](https://jira.stsci.edu/secure/ViewProfile.jspa?name=kvolk):

I ran the same level3 image combination with 4 cal images in version 1.14.0 and 1.15.0 while looking at "top" to try to see if there was a difference in memory usage between the two cases.  I did not see a clear difference in the peak memory usage recorded since the peak was 0.3% and so the precision of the value was low.  However, the run was about 16% faster in 1.15.0 than in 1.14.0 on witserv4 when I was the only user that was active.  So it does appear that the changes improve the performance.
